libcnb-cargo/src/lib.rs

```diff
@@ -0,0 +1,205 @@
+        .and_then(|file_contents| {
+            toml::from_str(&file_contents).map_err(BuildpackDataError::DeserializationError)
+        })
+        .map(|buildpack_toml| BuildpackData {
+            buildpack_toml_path,
+            buildpack_toml,
+        })
+}
+
+#[derive(Debug)]
+pub enum BuildpackDataError {
+    IoError(std::io::Error),
+    DeserializationError(toml::de::Error),
+}
+
+pub struct BuildpackData<BM> {
+    pub buildpack_toml_path: PathBuf,
+    pub buildpack_toml: BuildpackToml<BM>,
+}
+
+/// Assembles a buildpack tarball and writes it to the given destination path.
+///
+/// Assembly of the tarball follows the constraints set by the libcnb framework. For example,
+/// the buildpack binary is only stored once and symlinks are used to refer to it when the CNB
+/// spec requires different file(name)s.
+///
+/// This function will not validate if the buildpack descriptor at the given path is valid and will
+/// use it as-is.
+///
+/// # Errors
+/// Will return `Err` if the tarball could not be assembled successfully.
+pub fn assemble_buildpack_tarball(
+    destination_path: impl AsRef<Path>,
+    buildpack_toml_path: impl AsRef<Path>,
+    buildpack_binary_path: impl AsRef<Path>,
+) -> std::io::Result<()> {
+    let destination_file = fs::File::create(destination_path.as_ref())?;
+    let mut buildpack_toml_file = fs::File::open(buildpack_toml_path.as_ref())?;
+    let mut buildpack_binary_file = fs::File::open(buildpack_binary_path.as_ref())?;
+
+    let mut tar_builder =
+        tar::Builder::new(GzEncoder::new(destination_file, Compression::default()));
+
+    tar_builder.append_file("buildpack.toml", &mut buildpack_toml_file)?;
+    tar_builder.append_file("bin/build", &mut buildpack_binary_file)?;
+
+    // Build a symlink header to link `bin/detect` to `bin/build`
+    let mut header = Header::new_gnu();
+    header.set_entry_type(EntryType::Symlink);
+    header.set_path("bin/detect")?;
+    header.set_link_name("build")?;
+    header.set_size(0);
+    header.set_mtime(
+        std::time::SystemTime::now()
+            .duration_since(std::time::UNIX_EPOCH)
+            .unwrap_or(std::time::Duration::ZERO)
+            .as_secs(),
+    );
+    header.set_cksum();
+
+    tar_builder.append(&header, &[][..])?;
+
+    tar_builder.into_inner()?.finish()?.flush()
+}
+
+/// Construct a good default filename for a buildpack tarball.
+///
+/// It uses the given buildpack metadata and cargo profile to construct a good default name for the
+/// buildpack tarball. This function ensures the resulting filename is valid and does not contain
+/// problematic characters such as `/`.
+pub fn default_buildpack_tarball_filename<BM>(
+    buildpack_toml: &BuildpackToml<BM>,
+    cargo_profile: CargoProfile,
+) -> String {
+    format!(
+        "{}_{}_{}.tar.gz",
+        buildpack_toml.buildpack.id.replace("/", "_"),
+        buildpack_toml.buildpack.version,
+        match cargo_profile {
```

Comment:
should we just implement the `Display` trait on `CargoProfile` or have a `target_dir` fn?

Reply:
I don't think we should implement `Display` for `CargoProfile`. Looking at this code, you can see that we map the profile to different strings throughout. For example, Cargo puts the binaries created with the `dev` profile in a directory called `debug`.